src/index.ts

```diff
@@ -73,22 +135,110 @@ class AsyncSource<T> {
         this.responseData = null;
         this.lastRequestId = 0;
+        this.removeCachedData();
+    }
+
+    private async syncCache(args: Array<any>) {
+        this.updateCacheKey(args);
+
+        const cachedData = await this.getCachedData();
+        if (cachedData) {
+            this.responseData = cachedData;
+            this.isFetchedData = true;
+            this.isRequestPending = false;
+        }
+    }
+
+    private updateCacheKey(requestParams: Array<any>) {
+        const baseCacheKey = `${AsyncSource.cachePrefix}-${this.requestCacheKey}`;
+
+        const isRequestParamsExist = Boolean(requestParams.length);
+
+        let requestParamsHash;
+
+        if (isRequestParamsExist) {
+            try {
+                requestParamsHash = JSON.stringify(requestParams);
+            } catch (error) {
+                console.warn({ message: `Request params saving error cacheKey:${baseCacheKey}`, params: requestParams, error });
+            }
+        }
+
+        this.cacheKey = requestParamsHash ? `${baseCacheKey}-${requestParamsHash}` : baseCacheKey;
     }
 
-    // Core request method
-    private async request(args: Array<any>, successHandler?: ((response: T) => void), isImmediate?: boolean) {
+    private removeCachedData() {
+        try {
+            if (this.cacheKey) {
+                this.cacheStorage?.removeItem?.(this.cacheKey);
+            }
+        } catch (error) {
+            console.warn({ message: `Cache removing error cacheKey:${this.cacheKey}`, error });
+        }
+    }
+
+    private async getCachedData(): Promise<PromiseResult<T> | null> {
+        if (!this.cacheKey) return null;
+
+        try {
+            const storedData = await this.cacheStorage?.getItem?.(this.cacheKey);
+
+            if (storedData) {
+                const { data, timestamp } = JSON.parse(storedData);
+                const isExpired = Date.now() - timestamp > this.cacheTime;
+                return isExpired ? null : (data as PromiseResult<T>);
+            }
+
+            return null;
+        } catch (error) {
+            console.warn({ message: `Cache getting error cacheKey:${this.cacheKey}`, error });
+            return null;
+        }
+    }
+
+    private setCachedData(data: T): void {
+        if (!this.cacheKey) return;
+
+        const cacheValue = {
+            data,
+            timestamp: Date.now()
+        };
+
+        try {
+            this.cacheStorage?.setItem?.(this.cacheKey, JSON.stringify(cacheValue));
+        } catch (error) {
+            console.warn({ message: `Cache saving error cacheKey:${this.cacheKey}`, error });
+        }
+    }
+
+    private async request(args: Array<any>, successHandler?: (response: T) => void, isImmediate?: boolean): Promise<void> {
         this.isRequestPending = true;
 
         const requestId = await this.createRequestId(isImmediate);
         if (!this.isLastRequest(requestId)) return;
 
+        if (this.isCacheEnabled) {
```

Comment:
```
if (this.isCacheEnabled && this.isLastRequest(requestId)) {
```
It seems it should be like this.

Reply:
`if (!this.isLastRequest(requestId)) return;` but we have this? it's the same check i think